### PR TITLE
Fix: Stripe Elements are now mounted in same-page checkout setup

### PR DESCRIPTION
**Pull request by unixtam**
As per #511, I'm re-submitting this fix and describing in more detail why this change is warranted.

It can be helpful to have product selection and checkout on the same page. Several solutions exist.

One way to make this work is by loading a hidden checkout form template on page load and then populating it with payment options via AJAX fragments (`.woocommerce-checkout-payment` fragment in response to `update_order_review` action).

Currently, `assets/js/stripe.js` checks for the existence of `#stripe-card-element` before the Stripe Elements are created. This code runs on page load, at which point the hidden checkout form is not yet populated with payment options. The result is that Stripe Elements are not being created and credit card information cannot be entered once the checkout form is made visible.

To solve this, I have pushed back the check for `#stripe-card-element` to before the Elements are mounted. Now the Elements are created on page load regardless of the presence of `#stripe-card-element` and mounted from the code hooked to the `updated_checkout` event in line 121. At that point the payment options are present in the DOM and the Elements are mounted as expected.

### CODE REVIEW

File: assets/js/stripe.js [hunk 1]
```diff
@@ -40,6 +40,9 @@ jQuery( function( $ ) {
 		},
 
 		mountElements: function() {
+			if ( !$( '#stripe-card-element' ).length ) {
```
Comment (roykho) [inline assets/js/stripe.js]: Coding standards: Missing space after `!`

**Comment by unixtam** [comment]
> Have you tested this to see if it breaks Add payment page? Also Change payment method page for Subscriptions? What about Pay for Order page?

I have now, it's all fine. I've noticed the margin for Stripe radio buttons seemed slightly off on the Pay for Order page, but that is unrelated to my changes.

File: assets/js/stripe.js (hunk 1 above)
Comment (unixtam) [reply assets/js/stripe.js]: fixed

**Comment by roykho** [comment]
Hi @unixtam do you have a way for me to test this? Are you using a plugin? Perhaps "One page Checkout" plugin or something?

**Comment by denzel100** [comment]
It's a lot of custom, integrated and site-specific code; not sure how to make it available for you to properly test it. However, the One Page Checkout plugin might run into the same problem, so that's something you could test. If you do test, make sure the cart is empty on page load.

Screenshots for comparison:

Before PR:

elements_bug

After PR:

elements_fixed

**Comment by unixtam** [comment]
It's a lot of custom, integrated and site-specific code; not sure how to make it available for you to properly test it. However, the One Page Checkout plugin might run into the same problem, so that's something you could test. If you do test, make sure the cart is empty on page load.

Screenshots for comparison:

Before PR:

elements_bug

After PR:

elements_fixed

**Comment by unixtam** [comment]
I've had a chance to test WooCommerce One Page Checkout (WC OPC) in particular. In many cases it's unaffected because it outputs the payment options on page load, so that `#stripe-card-element` is present. You can reproduce the issue for testing purposes by commenting out the code inside `PP_One_Page_Checkout::maybe_toggle_cart_needs_payment()` in `woocommerce-one-page-checkout.php`. Then create an OPC product, open the cart page, make sure the cart is empty, open the product page, add the product to the order, find CC Elements are not mounted.

But WC OPC *is* affected when used together with plugins that limit payment options on page load, such as "WooCommerce - Country Based Payments". If the default billing country (saved or geolocated) is not within the list of countries for which credit card payment is accepted, this payment option is not available on page load but dynamically added via AJAX upon selecting a different billing country. Since `#stripe-card-element` is absent on page load, the CC Elements are broken (and fixed by this PR).

This PR fixes the issue for all same-page checkout solutions, regardless of whether the Stripe CC payment option is present on page load.

**Comment by roykho** [comment]
Thanks for the PR

**Pull request merged**